src/coreclr/jit/indirectcalltransformer.cpp

```diff
@@ -540,28 +540,39 @@ class IndirectCallTransformer
             checkBlock             = currBlock;
             checkBlock->bbJumpKind = BBJ_COND;
 
-            CallArg* thisArg  = origCall->gtArgs.GetThisArg();
-            GenTree* thisTree = thisArg->GetNode();
-
-            // Create temp for this if the tree is costly.
-            if (thisTree->IsLocal())
-            {
-                thisTree = compiler->gtCloneExpr(thisTree);
-            }
-            else
+            // The guard is going to dereference 'this'. This needs to happen
+            // after arguments have been evaluated if 'this' may be null, so we
+            // have to spill all arguments with side effects.
+            for (CallArg& arg : origCall->gtArgs.Args())
             {
-                const unsigned thisTempNum = compiler->lvaGrabTemp(true DEBUGARG("guarded devirt this temp"));
-                GenTree*       asgTree     = compiler->gtNewTempAssign(thisTempNum, thisTree);
-                Statement*     asgStmt     = compiler->fgNewStmtFromTree(asgTree, stmt->GetDebugInfo());
-                compiler->fgInsertStmtAtEnd(checkBlock, asgStmt);
+                bool     spill   = false;
+                GenTree* argNode = arg.GetNode();
 
-                thisTree = compiler->gtNewLclvNode(thisTempNum, TYP_REF);
+                if (arg.GetWellKnownArg() == WellKnownArg::ThisPointer)
+                {
+                    // Guard is going to access this arg, spill it if it is
+                    // complex, regardless of side effects.
+                    spill = !argNode->IsLocal();
+                }
+                else
+                {
+                    spill = (argNode->gtFlags & GTF_SIDE_EFFECT) != 0;
```

Comment:
This isn't sufficient. We also have to spill all args with `GTF_GLOB_REF` before we spill these. Unfortunately I don't think `GTF_GLOB_REF` is accurate at this point.

Reply:
`((argNode->gtFlags & GTF_GLOB_REF) != 0) || gtHasLocalsWithAddrOp(argNode)`?

Reply:
Yeah, unfortunately I suppose that's going to be on the expensive side TP and CQ wise, but I'll try it out.